Testing

johnwhitton opened the pull request:
This pull requests is to enhance testing of smart contracts.
Specifically around tokens and Registrars.

polymorpher commented:
Can you move docs stuff to a separate PR? Also let's add code/build/* to .gitignore

polymorpher commented on code/test/checkUtil.js:
```diff
@@ -0,0 +1,45 @@
+// get OneWallet state
+const getONEWalletState = async (wallet) => {
```
If we want to compare the state against something (especially javascript object) we might want to parse each of the state variables instead of using object returned by web3/truffle. But for informational purposes this looks fine for now

johnwhitton replied:
for an example see [here](https://github.com/polymorpher/one-wallet/blob/master/code/lib/api/index.js#L262)

johnwhitton replied:
@polymorpher Thanks for the example :pray:
Have updated `getONEWalletState` and `checkONEWallet` to use plain JS objects in the [latest commit](https://github.com/1wallet-dev/1wallet/commit/d6a4e8e11dcd7f82beba8ff54a97c68882986463)

@polymorpher could you review specifically the following conversions for nonce and lastOperationTime
```
  // const nonce = new BN(await wallet.getNonce())
  const nonce = (await wallet.getNonce()).toNumber()
  const lastOperationTime = new BN(await wallet.lastOperationTime())
  ```
  The goal is to be able to easily compare the state values agains those returned by the function call and update the state to the new values where possible.

johnwhitton replied:
Updating and will push shortly
Currently for all numbers we're converting them to BN and then toNumber()
e.g.
`const nonce = new BN(await wallet.getNonce()).toNumber()`
except for SpendingStates which are stored in wei and if we use this approach we get an exception parsing
`Failed to parse walletSpendingState: Error: Number can only safely store up to 53 bits`
as a workaround we are converting these back to ETH first using
`spendingLimit: new BN(unit.fromWei(walletSpendingState[0], 'ether')).toNumber(),`
`spendingAmount: new BN(unit.fromWei(walletSpendingState[1], 'ether')).toNumber(),`
`highestSpendingLimit: new BN(unit.fromWei(walletSpendingState[5], 'ether')).toNumber(),`

below is an example of the state returned
```
{
    "address": "0x2F62d7CF38576709828537D9E7e5DebE1f6E4750",
    "identificationKey": "0xd9a9eb4f14cde1ada8c3b04b063b0da893ae137e1ed8fdd52310908f44bb5740fd38f62e32072fe93cd1e763dd8dd95ed6c8eadc5f49fd5d57c68671266137e0",
    "identificationKeys": {
        "0": {
            "key": "0"
        }
    },
    "forwardAddress": "0x0000000000000000000000000000000000000000",
    "info": {
        "root": "0xe670b093536fdd441c35c9811bd2d29ee441b945bf9582804596803b11e9977e",
        "height": 5,
        "interval": 30,
        "t0": 54921498,
        "lifespan": 12,
        "maxOperationsPerInterval": 1,
        "recoveryAddress": "0x2B0178e0F3fBC9c0f43825FdC14845dcEBb50F3F",
        "extra": 0
    },
    "oldInfo": {},
    "innerCores": {
        "0": {
            "root": "0xe0d92235fc0f9dac939d213214dde318baaa5a1fb89bc48924bfea2ff5fb02c7",
            "height": 2,
            "interval": 180,
            "t0": 9153583,
            "lifespan": 2,
            "maxOperationsPerInterval": 1
        },
        "1": {
            "root": "0x1dbeeb8c2d1ef9f31db9deebac0ace41d16a4c13bbeddd0d09e51e25b862fc12",
            "height": 2,
            "interval": 180,
            "t0": 9153583,
            "lifespan": 2,
            "maxOperationsPerInterval": 1
        },
        "2": {
            "root": "0x4d2faff76619f7679f56f9abfd12b020faddbb1d193584db8710bb106a22f717",
            "height": 2,
            "interval": 180,
            "t0": 9153583,
            "lifespan": 2,
            "maxOperationsPerInterval": 1
        },
        "3": {
            "root": "0x7ea9df9143ecdaf59348fa4fab1c50ef88e9a56b6bdeb39ae3885c7521524ade",
            "height": 2,
            "interval": 180,
            "t0": 9153583,
            "lifespan": 2,
            "maxOperationsPerInterval": 1
        },
        "4": {
            "root": "0xc41607660b9658c4448b537619cefc1d0117237a13c44bdee64a06548ad7e13f",
            "height": 2,
            "interval": 180,
            "t0": 9153583,
            "lifespan": 2,
            "maxOperationsPerInterval": 1
        },
        "5": {
            "root": "0xab7a676d7bd2de67100b05dd1bc478b431dbe52f905ded264e60e4a93fe7be40",
            "height": 2,
            "interval": 180,
            "t0": 9153583,
            "lifespan": 2,
            "maxOperationsPerInterval": 1
        }
    },
    "rootKey": "0xe670b093536fdd441c35c9811bd2d29ee441b945bf9582804596803b11e9977e",
    "version": {
        "majorVersion": 15,
        "minorVersion": 1
    },
    "spendingState": {
        "spendingLimit": 1,
        "spendingAmount": 0,
        "lastSpendingInterval": 0,
        "spendingInterval": 86400,
        "lastLimitAdjustmentTime": 0,
        "highestSpendingLimit": 1
    },
    "nonce": 1,
    "lastOperationTime": 1647645262,
    "allCommits": {
        "commitHashArray": [
            "0x1f9676488ca85db123df4f1c6e6556ee8a8bf8a8febfd1282535f0299bab634d"
        ],
        "paramHashArray": [
            "0x3ded2f8f86e120dd54f35729ce6841bba4e16cb218bbdec30964a3db53c7c2a2"
        ],
        "veriFicationHashArray": [
            "0x3cccb9b5b60f85486f2d656cefe145d302b25910ccfc8180cea91b3c5e55e12e"
        ],
        "timestampArray": [
            "6235124d"
        ],
        "completedArray": [
            true
        ]
    },
    "trackedTokens": {
        "tokenTypeArray": [
            "0"
        ],
        "contractAddressArray": [
            "0xCb36071971a20714d982f75f44ea95E13BB87e60"
        ],
        "tokenIdArray": [
            "0"
        ]
    },
    "backlinks": {}
}
```

johnwhitton replied:
Also have changed the `assetTransfer` logic to store the `oldState` and update the `currentState`
and changed `checkONEWallet` to be readOnly simply comparing the `oldState` vs `newState` rather than retrieving and updating the currentState.

This role of testing is to 
1. Identify differences (checking new values are correct when they are deterministic)
2. Update the oldState for these values to the currentState values
3. check that there are no other differences in State (i.e. updated oldState = currentState)

Below are two approaches for doing this 
Approach A - use oldState and CurrentState exclusively.
This puts a heavier reliance that the values retrieved from `getONEWalletState` are accurate
example
```
    // Alice Items that have changed - nonce, spendingState, lastOperationTime, Commits
    assert.notEqual(alice.currentState.nonce, alice.oldState.nonce, 'alice wallet.nonce should have been changed')
    assert.equal(alice.currentState.nonce, alice.oldState.nonce + 1, 'alice wallet.nonce should have been changed')
    alice.oldState.nonce = alice.currentState.nonce
    // spendingState
    assert.equal(alice.currentState.spendingState.spentAmount, (ONE_CENT / 2).toString(), 'alice wallet.spentAmount should have been changed')
    alice.oldState.spendingState.spentAmount = alice.currentState.spendingState.spentAmount
    assert.notEqual(alice.currentState.spendingState.lastSpendingInterval, '0', 'alice wallet.spentAmount should have been changed')
    alice.oldState.spendingState.lastSpendingInterval = alice.currentState.spendingState.lastSpendingInterval
    // lastOperationTime
    assert.notStrictEqual(alice.currentState.lastOperationTime, alice.oldState.lastOperationTime, 'alice wallet.lastOperationTime should have been updated')
    alice.oldState.lastOperationTime = alice.currentState.lastOperationTime
    // commits
    assert.notDeepEqual(alice.currentState.allCommits, alice.oldState.allCommits, 'alice wallet.allCommits should have been updated')
    alice.oldState.allCommits = alice.currentState.allCommits
    // check alice
    await CheckUtil.checkONEWallet(alice)
```
Approach B - retrieve the values again in the test suite and use them wherever possible
This places an additional validation layer on top of `getONEWalletState` but can add complexity to the writing of the harnesses. **Also note: there is not a good solution for complex objects like `allCommits`**
example
```
   // Alice Items that have changed - nonce, spendingState, lastOperationTime, Commits
    const nonce = await await alice.wallet.getNonce()
    assert.notEqual(nonce, alice.oldState.nonce, 'alice wallet.nonce should have been changed')
    assert.equal(nonce.toNumber(), alice.oldState.nonce + 1, 'alice wallet.nonce should have been changed')
    alice.oldState.nonce = nonce.toNumber()
    // spendingState
    const spendingState = await alice.wallet.getSpendingState()
    assert.equal(spendingState.spentAmount, (ONE_CENT / 2).toString(), 'alice wallet.spentAmount should have been changed')
    alice.oldState.spendingState.spentAmount = spendingState.spentAmount
    assert.notEqual(spendingState.lastSpendingInterval, '0', 'alice wallet.spentAmount should have been changed')
    alice.oldState.spendingState.lastSpendingInterval = spendingState.lastSpendingInterval
    // lastOperationTime
    const lastOperationTime = await alice.wallet.lastOperationTime()
    assert.notStrictEqual(lastOperationTime, alice.oldState.lastOperationTime, 'alice wallet.lastOperationTime should have been updated')
    alice.oldState.lastOperationTime = lastOperationTime.toNumber()
    // commits
    const allCommits = await alice.wallet.getAllCommits()
    assert.notDeepEqual(allCommits, alice.oldState.allCommits, 'alice wallet.allCommits should have been updated')
    alice.oldState.allCommits = alice.currentState.allCommits
```

johnwhitton replied:
anything declared as arrays in IONEWallet should be arrays not objects
e.g. identificationKeys

johnwhitton replied:
Updated to arrays where appropriate latest example of state below
```
{
    "address": "0x219764E1eDF3F59952880EC979d6292f285A2Eed",
    "identificationKey": "0x0b4df4159e9bb0a059a6ddd8d9e3e8991c4607ce58e98e06fdcbbfc06ea3c92687840b86f07041d37169b05c35c4cdfff117be320e87170fb5834d6c8cdc729c",
    "identificationKeys": [
        "0"
    ],
    "forwardAddress": "0x0000000000000000000000000000000000000000",
    "info": {
        "root": "0x8196b1243c545ca4f1660a2865da8a67dc0b02c34629a7eba84a2d5849d57835",
        "height": 5,
        "interval": 30,
        "t0": 54924540,
        "lifespan": 12,
        "maxOperationsPerInterval": 1,
        "recoveryAddress": "0xC6A240e846683cA22F1200e9bc22ef396aA983c0",
        "extra": 0
    },
    "oldInfo": [],
    "innerCores": [
        {
            "root": "0x5d3a7b7308317ad954cebcf7841f6def609c5a3c6638b2d6f263d0deafeb6220",
            "height": 2,
            "interval": 180,
            "t0": 9154090,
            "lifespan": 2,
            "maxOperationsPerInterval": 1
        },
        {
            "root": "0x6f3abce8d5e9471d0160769afefab26869989f006ac7da94c234b539236cf104",
            "height": 2,
            "interval": 180,
            "t0": 9154090,
            "lifespan": 2,
            "maxOperationsPerInterval": 1
        },
        {
            "root": "0x9f408eadc7a15456243716c6b031260c677eb4337f2f27e1c5027c88cac1db2f",
            "height": 2,
            "interval": 180,
            "t0": 9154090,
            "lifespan": 2,
            "maxOperationsPerInterval": 1
        },
        {
            "root": "0xd0fb7aec8c16167246f7bd65945094666b09231f080c2a8af7a0960be27d0eed",
            "height": 2,
            "interval": 180,
            "t0": 9154090,
            "lifespan": 2,
            "maxOperationsPerInterval": 1
        },
        {
            "root": "0x7182904842fc242162431dedb170859b590e46abb3e1e9ce2bdee1f06ac07fdd",
            "height": 2,
            "interval": 180,
            "t0": 9154090,
            "lifespan": 2,
            "maxOperationsPerInterval": 1
        },
        {
            "root": "0xb8b19a862db699f2832af5a62a61cade4caf7f4e2817b1b4915158a0c3701546",
            "height": 2,
            "interval": 180,
            "t0": 9154090,
            "lifespan": 2,
            "maxOperationsPerInterval": 1
        }
    ],
    "rootKey": "0x8196b1243c545ca4f1660a2865da8a67dc0b02c34629a7eba84a2d5849d57835",
    "version": {
        "majorVersion": 15,
        "minorVersion": 1
    },
    "spendingState": {
        "spendingLimit": "1000000000000000000",
        "spentAmount": "5000000000000000",
        "lastSpendingInterval": "19071",
        "spendingInterval": "86400",
        "lastLimitAdjustmentTime": "0",
        "highestSpendingLimit": "1000000000000000000"
    },
    "nonce": 1,
    "lastOperationTime": 1647736528,
    "allCommits": {
        "commitHashArray": [
            "0x7be9bee01d72f71343f19c1c1ff39a9bd133f8d0704c891a1141184b09164c1d"
        ],
        "paramHashArray": [
            "0x59c3caaeb509f2c7ea7e6b343ea6bc9f855163f185a9279e659fb9c0078811de"
        ],
        "veriFicationHashArray": [
            "0x215feaf23e1ce1c2759fa59842d467eb3760ae10f2b041e2cccf35b038b9e871"
        ],
        "timestampArray": [
            "623676d0"
        ],
        "completedArray": [
            true
        ]
    },
    "trackedTokens": {
        "tokenTypeArray": [
            "0"
        ],
        "contractAddressArray": [
            "0x69bF30aC4B8dE7bD0E3C6C56aF45184082D803F7"
        ],
        "tokenIdArray": [
            "0"
        ]
    },
    "backlinks": []
}
```

polymorpher replied:
Okay, some details can be changed later to be consistent with what's stored on the contract and expected at the client

polymorpher commented on code/test/util.js:
```diff
@@ -257,7 +255,22 @@ const printInnerTrees = ({ Debugger, innerTrees }) => {
 }
 
+// Create Accounts Alice Bob and Carol and fund them
+const createAccounts = async (initArgs) => {
```
Create brand new accounts here? If the provider is instantiated with a private key, web3.eth.getAccounts() would only give you one account. Even if it instantiated by mnemonics (default setting), accounts at 0, 1, 2 often have balance and account 0 is used for execution by default.

johnwhitton replied:
done 
Below was the alternate strategy which allowed each wallet to be assoicated with a resortAddress which could sign transactions. Useful if we want to test specific accounts
It was leaving account[0] as the "deployer" and having accounts 1 though 9 with 1000 native tokens

#### GANACHE TESTING

Note: The mnemonic and private keys shown are for testing only and should never be used in production. There is nothing special about this mnemonic you can generate your own using a tool such as https://iancoleman.io/bip39/ and then update your `.env` file with the private keys show when starting ganache.

For smart contract testing you should also install [ganache](https://trufflesuite.com/docs/ganache/quickstart.html). Once installed you should run ganache whenever using the `dev` or `ganache` networks. These networks have been configured to use port 7545 as such you can run ganache using the command 
```
ganache --port 7545 -m 'filter group there hunt fitness junior ghost park route jar entire clown allow rifle meadow'
```
by providing the mnemonic (using `-m`) you are guaranteed to receive the same mnemonic keys which you can configure in your `.env` file see `.env.sample` for an example.

johnwhitton replied:
This has now been modified to create new accounts using a unique identifier
TT - Two Character Test Identifier (one per test file)
XX - Sub Test one per test 
99 - Wallet Identifier with test
Account creation logic
`   let alice = await CheckUtil.makeWallet('TT-ERC20-1', accounts[0], EFFECTIVE_TIME)`
Wallet Creation Function - note `effectiveTime` (the wallets effective creation time) is defined in the test and passed.
```
// makeWallet uses an index and unlocked web3.eth.account and creates and funds a ONEwallet
const makeWallet = async (salt, deployer, effectiveTime) => {
  const lastResortAccount = web3.eth.accounts.create()
  const { wallet, seed, hseed, root, client: { layers } } = await TestUtil.createWallet({
    salt: new BN(ONEUtil.keccak(salt)),
    effectiveTime: effectiveTime,
    duration: DURATION,
    maxOperationsPerInterval: SLOT_SIZE,
    lastResortAddress: lastResortAccount.address,
    spendingLimit: ONE_ETH
  })
  // Fund wallet
  await web3.eth.sendTransaction({
    from: deployer,
    to: wallet.address,
    value: HALF_ETH
  })
  const currentState = await getONEWalletState(wallet)
  return { wallet, seed, hseed, root, layers, lastResortAddress: lastResortAccount.address, currentState }
}
```

polymorpher commented on code/test/tokens.js:
```diff
@@ -0,0 +1,472 @@
+const makeWallet = async (accountIndex, lastResortAddress) => {
+  // create wallet
+  const EFFECTIVE_TIME = Math.floor(Date.now() / INTERVAL / 6) * INTERVAL * 6 - DURATION / 2
+  const TEN_ETH = unit.toWei('10', 'ether')
```
I suggest using a small amount for testing in cases that some tests won't work with snapshot-revert pattern

polymorpher commented on code/test/tokens.js:
```diff
@@ -0,0 +1,472 @@
+  await web3.eth.sendTransaction({
+    from: lastResortAddress,
+    to: wallet.address,
+    value: TEN_ETH
```
I suggest using a small amount for testing in cases that some tests won't work with snapshot-revert pattern

johnwhitton replied:
Agreed - will update

johnwhitton replied:
Done this now uses 0.5 ETH for funding with the transfers currently being for 0.01 or 0.005 Ether

polymorpher commented on code/test/tokens.js:
```diff
@@ -0,0 +1,472 @@
+  const uris1155 = ['ipfs://test1155/0', 'ipfs://test1155/1', 'ipfs://test1155/2', 'ipfs://test1155/3', 'ipfs://test1155/4', 'ipfs://test1155/5', 'ipfs://test1155/6', 'ipfs://test1155/7', 'ipfs://test1155/8', 'ipfs://test1155/9']
+  const testerc1155 = await TESTERC1155.new(tids1155, amounts1155, uris1155, { from: owner })
+  // create a TokenTracker
+  const TOKENTRACKER = artifacts.require('TokenTracker')
```
same

polymorpher commented on code/test/tokens.js:
```diff
@@ -0,0 +1,472 @@
+}
+
+// tokenTransfer commits and reveals a wallet transaction
+const tokenTransfer = async ({ wallet, operationType, tokenType, contractAddress, tokenId, dest, amount }) => {
```
Let's call it assetTransfer, to cover both native asset (ONE) from tokens, and in general distinguish the two

johnwhitton replied:
done

polymorpher commented on code/test/tokens.js:
```diff
@@ -0,0 +1,472 @@
+          revealParams = { operationType, tokenType, contractAddress, dest, amount }
+          break
+        case ONEConstants.TokenType.ERC721:
+          commitParams = { operationType, tokenType, contractAddress, tokenId, dest, amount }
```
amount would be ignored for ERC721

johnwhitton replied:
done

polymorpher commented on code/test/tokens.js:
```diff
@@ -0,0 +1,472 @@
+  // Transfer Native Token to external wallet
+  it('Wallet_CommitReveal: Native Token Transfer must commit and reveal successfully', async () => {
+    // Create Wallets and tokens
+    const alice = await makeWallet(1, accounts[1])
```
I suggest using a unique number specific to the test (e.g. hash of the name, then converted to BN) for salt. The salt will determine the wallet address, so a unique value ensures it doesn't collide with any existing wallet

johnwhitton replied:
Agreed - will implement

johnwhitton replied:
This has now been modified to create new accounts using a unique identifier
TT - Two Character Test Identifier (one per test file)
XX - Sub Test one per test 
99 - Wallet Identifier with test
Account creation logic
`   let alice = await CheckUtil.makeWallet('TT-ERC20-1', accounts[0], EFFECTIVE_TIME)`
Wallet Creation Function - note `effectiveTime` (the wallets effective creation time) is defined in the test and passed.
```
// makeWallet uses an index and unlocked web3.eth.account and creates and funds a ONEwallet
const makeWallet = async (salt, deployer, effectiveTime) => {
  const lastResortAccount = web3.eth.accounts.create()
  const { wallet, seed, hseed, root, client: { layers } } = await TestUtil.createWallet({
    salt: new BN(ONEUtil.keccak(salt)),
    effectiveTime: effectiveTime,
    duration: DURATION,
    maxOperationsPerInterval: SLOT_SIZE,
    lastResortAddress: lastResortAccount.address,
    spendingLimit: ONE_ETH
  })
  // Fund wallet
  await web3.eth.sendTransaction({
    from: deployer,
    to: wallet.address,
    value: HALF_ETH
  })
  const currentState = await getONEWalletState(wallet)
  return { wallet, seed, hseed, root, layers, lastResortAddress: lastResortAccount.address, currentState }
}
```

polymorpher commented on code/test/tokens.js:
```diff
@@ -0,0 +1,472 @@
+  })
+
+  // ERC20 Decimals 9 Testing (Transfer, Mint, Track, SpendingLimit)
+  it('Wallet_CommitReveal: ERC20-9(Transfer, Mint, Track) must commit and reveal successfully', async () => {
```
The purpose of Decimals-9 test contract is primarily to test whether tokens with less decimals (e.g. USDT) are displayed correctly in the UI (in terms of amount). The focus would be on the conversion functions (ONEUtil.toFraction, etc.). So a separate test at the smart contract level is probably not needed, unless it involves testing formatting and value conversion functions

johnwhitton replied:
Agreed - removed

polymorpher commented on code/test/tokens.js:
```diff
@@ -0,0 +1,472 @@
+    // commits
+    assert.notDeepEqual(await alice.wallet.getAllCommits(), alice.state.allCommits, 'alice wallet.allCommits should have been updated')
+    alice.state.allCommits = await alice.wallet.getAllCommits()
+    // tracked tokens
```
It should check whether this particular ERC20 token is now tracked by Alice, but not Bob

johnwhitton replied:
Agreed - will implement.

polymorpher commented on code/test/tokens.js:
```diff
@@ -0,0 +1,472 @@
+    assert.equal(20, await testerc1155.balanceOf(alice.lastResortAddress, 8), 'Alice.lastResortAddress owns 20 of token 8')
+    // transfer ERC721 tokens from alice.lastResortAddress (which owns the tokens) to alices wallet
+    // TODO review the bytes value we are passing in safeTransferFrom (currently using ONEUtil.hexStringToBytes('5') )
+    await testerc1155.safeTransferFrom(alice.lastResortAddress, alice.wallet.address, 8, 8, ONEUtil.hexStringToBytes('5'), { from: alice.lastResortAddress })
```
5 is not really a hexString

johnwhitton replied:
Have updated to use a constant for a dummy string
`const FIVE_HEX = ONEUtil.hexString('5')` which is `0x05`

polymorpher commented:
I suggest using web3.eth.accounts.create and keep minimal assumption on provider/accounts set in environment (e.g. using a single account / private key, so we can test not only in local environments)

johnwhitton commented on code/test/tokens.js:
```diff
@@ -0,0 +1,472 @@
+})
+
+// makeWallet uses an index and unlocked web3.eth.account and creates and funds a ONEwallet
+const makeWallet = async (accountIndex, lastResortAddress) => {
```
done
They are currently in `checkUtil.js` (but can move them to `util.js` if we feel that we are happy with them and want to consolidate the utilty files.

johnwhitton replied:
I'm closing this for now - have moved the utility functions into checkUtil.

johnwhitton commented on code/test/tokens.js:
```diff
@@ -0,0 +1,472 @@
+// makeTokens makes test ERC20, ERC20Decimals9, ERC721, ERC1155
+const makeTokens = async (owner) => {
+  // create an ERC20
+  const TESTERC20 = artifacts.require('TestERC20')
```
done

johnwhitton commented on code/test/tokens.js:
```diff
@@ -0,0 +1,472 @@
+  })
+
+  // Transfer Native Token to external wallet
+  it('Wallet_CommitReveal: Native Token Transfer must commit and reveal successfully', async () => {
```
updated

johnwhitton commented on code/test/tokens.js:
```diff
@@ -0,0 +1,472 @@
+  const testerc1155 = await TESTERC1155.new(tids1155, amounts1155, uris1155, { from: owner })
+  // create a TokenTracker
+  const TOKENTRACKER = artifacts.require('TokenTracker')
+  const tokentracker = await TOKENTRACKER.new()
```
Agreed and removed.

johnwhitton commented on code/test/tokens.js:
```diff
@@ -0,0 +1,472 @@
+    // alice.state.nonce = parseInt(alice.state.nonce) + 1
+    alice.state.nonce = await alice.wallet.getNonce()
+    // spendingState
+    assert.notStrictEqual(await alice.wallet.getSpendingState(), alice.state.spendingState, 'alice spendingState should have been changed')
```
Have updated `getONEWalletState` and `checkONEWallet` to use plain JS objects in the [latest commit](https://github.com/1wallet-dev/1wallet/commit/d6a4e8e11dcd7f82beba8ff54a97c68882986463)

Have also updated the testing
The goal is to be able to check for differences and where possible check against the value it should be updated to e.g. nonce should be increased by one.
After testing the difference we can set the state to the new value and check again.
Still working through assignments for `nonce` and `lastOperationTime`

johnwhitton commented on code/test/tokens.js:
```diff
@@ -0,0 +1,472 @@
+    const aliceInitialWalletBalance = await web3.eth.getBalance(alice.wallet.address)
+    assert.equal(TEN_ETH, aliceInitialWalletBalance, 'Alice Wallet initially has correct balance')
+    const bob = await makeWallet(2, accounts[2])
+    const { testerc721 } = await makeTokens(alice.lastResortAddress)
```
Agreed - will implement

johnwhitton replied:
Testing shows that alice is not tracking tokens we transfer two ERC721 tokens to alice and she transfers one of them to bob. But when checking her state. She is not tracking the ERC721 token that she still owns.

johnwhitton commented on code/test/tokens.js:
```diff
@@ -0,0 +1,472 @@
+    console.log(`alice.wallet.getTrackedTokens(): ${JSON.stringify(await alice.wallet.getTrackedTokens())}`)
+    console.log(`alice.state.trackedTokens: ${JSON.stringify(alice.state.trackedTokens)}`)
+    assert.notDeepEqual(await alice.wallet.getTrackedTokens(), alice.state.trackedTokens, 'alice.wallet.trackedTokens should have been updated')
+    alice.state.trackedTokens = await alice.wallet.getTrackedTokens()
```
agreed - will implement

johnwhitton commented on code/test/tokens.js:
```diff
@@ -0,0 +1,472 @@
+  })
+
+  // TokenTracker Testing (track, multitrack, getTrackedTokens, getBalance, recoverToken) also batch transactions
+  it('Wallet_CommitReveal: TokenTracker(token management) must commit and reveal successfully', async () => {
```
agreed - will implement

johnwhitton replied:
Have removed unused code.

johnwhitton commented:
Making progress - following are some outstanding items
- [ ] `Error: Returned error: VM Exception while processing transaction: revert Proof is incorrect` [commit](https://github.com/1wallet-dev/1wallet/commit/6b223d163b83a0a3dae56289b364601b57120524ttt)
- [ ] Parsing of values to plain JS Objects [commit](https://github.com/1wallet-dev/1wallet/commit/d6a4e8e11dcd7f82beba8ff54a97c68882986463)
- [ ] Token Tracking Tests
- [ ] Domain Manager Tests
- [ ] General Purpose Executor Tests (for multiple operation types)
  - [ ] Documenting and testing the `commitParams` and operationParams`
- [ ] Unique Account creation across tests

johnwhitton commented:
Summary of the intermittent  ` VM Exception while processing transaction: revert Proof is incorrect` issue

*Overview* 
When running multiple tests sequentially often get `revertProof is incorrect`. This is to be expected if we are executing multiple transactions on the same wallet which exceed the `maxOperationsPerInterval` currently this should be set to an `interval` of 30 seconds and `maxOperationsPerInterval` of 1. However we have tried two workarounds but am still gettin g these errors intermittently.

**Approach**
Code base is in the latest [testing branch](https://github.com/1wallet-dev/1wallet/tree/testing/code/test) and added in this [commit](https://github.com/1wallet-dev/1wallet/commit/6b223d163b83a0a3dae56289b364601b57120524)

**Approach 1: Wait 45 seconds between calls: currently implemented** 
This was done by exposing a [wait function](https://github.com/1wallet-dev/1wallet/blob/testing/code/test/util.js#L41) which is called between ONEWallet Transactions [e.g. here](https://github.com/1wallet-dev/1wallet/blob/testing/code/test/tokens.js#L298). Whilst not being a long term solution, this approach should mitigate the errors between ONEWallet operations. However we are still intermittently getting the above errors.

**Approach 2: Implement a `bumpTestTime` function to increase the effective test time**
This would work as follows

[bumpTestTime](https://github.com/1wallet-dev/1wallet/blob/testing/code/test/util.js#L58) would increase the `testEffectiveTime` and the networks time to a consistent future time (e.g. 45 seconds in the future) the code is below and it leverages `increaseTime` which calls `evm_increaseTime` and `evm_mine`

```
const bumpTestTime = (testEffectiveTime, bumpSeconds) => {
  testEffectiveTime = testEffectiveTime + (bumpSeconds * 1000)
  const chainBumpSeconds = testEffectiveTime - Date.now()
  increaseTime(chainBumpSeconds)
  console.log(`Date.now() .    : ${Date.now()}`)
  console.log(`testEffective   : ${testEffectiveTime}`)
  console.log(`chainBumpSeconds: ${chainBumpSeconds}`)
  return testEffectiveTime
}
```

[assetTransfer](https://github.com/1wallet-dev/1wallet/blob/testing/code/test/checkUtil.js#L64) is enhanced to accept an optional `testTime` which if passed is used in `ONEUtil.timeToIndex`

```
  if (testTime === undefined) { testTime = Date.now() }
  const effectiveTime = Math.floor(testTime / INTERVAL / 6) * INTERVAL * 6 - DURATION / 2
  const index = ONEUtil.timeToIndex({ effectiveTime })
  const eotp = await ONE.computeEOTP({ otp, hseed: wallet.hseed })
```
[tokens.js](https://github.com/1wallet-dev/1wallet/blob/testing/code/test/tokens.js) or any set of tests would be responsible for 
1. capturing the current time at the beginning of the test: `let testTime = Date.now()`
2.  increase the `testTime before each call to `assetTransfer`: `testTime = TestUtil.bumpTestTime(testTime, 45)`
3. add the `testTime` to the `assetTransfer` call e.g.
```
    await CheckUtil.assetTransfer(
      {
        wallet: alice,
        operationType: ONEConstants.OperationType.TRANSFER_TOKEN,
        tokenType: ONEConstants.TokenType.ERC1155,
        contractAddress: testerc1155.address,
        tokenId: 8,
        dest: bob.wallet.address,
        amount: 3,
        testTime
      }
```

**Sample Error Log when waiting **
```
  1) Contract: ONEWallet
       Wallet_CommitReveal: TokenTracker(token management) must commit and reveal successfully:
     Error: Returned error: VM Exception while processing transaction: revert Proof is incorrect
      at Object.commitReveal (test/util.js:273:44)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:95:5)
      at Object.assetTransfer (test/checkUtil.js:105:3)
      at Context.<anonymous> (test/tokens.js:317:5)
```

**Sample Error Log when bumping test time**
```
   1) Contract: ONEWallet
       Wallet_CommitReveal: TokenTracker(token management) must commit and reveal successfully:
     Error: Returned error: VM Exception while processing transaction: revert Proof is incorrect
      at Object.commitReveal (test/util.js:273:44)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:95:5)
      at Object.assetTransfer (test/checkUtil.js:105:3)
      at Context.<anonymous> (test/tokens.js:299:5)
 ```
 and
 ```
   1) Contract: ONEWallet
       Wallet_CommitReveal: TokenTracker(token management) must commit and reveal successfully:
     Error: Returned error: VM Exception while processing transaction: revert Time mismatch
      at Object.commitReveal (test/util.js:273:44)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:95:5)
      at Object.assetTransfer (test/checkUtil.js:105:3)
      at Context.<anonymous> (test/tokens.js:299:5)
 ```

johnwhitton commented on code/test/tokens.js:
```diff
@@ -0,0 +1,472 @@
+    assert.notDeepEqual(await alice.wallet.getAllCommits(), alice.state.allCommits, 'alice wallet.allCommits should have been updated')
+    alice.state.allCommits = await alice.wallet.getAllCommits()
+    // tracked tokens
+    console.log(`alice.wallet.getTrackedTokens(): ${JSON.stringify(await alice.wallet.getTrackedTokens())}`)
```
Checked and this is an issue. 
Will investigate and provide updates on root cause and resolution.

polymorpher commented on code/lib/onewallet.js:
```diff
@@ -222,6 +222,14 @@ const computeSetRecoveryAddressHash = ({ address }) => {
 }
 
+// dest, hex string
+const computeDestHash = ({ dest }) => {
```
To merge with computeSetRecoveryAddressHash

johnwhitton replied:
done - also modified `SetRecovey.jsx` to use `computeDestHash`

polymorpher replied:
where is it changed?

polymorpher commented on code/lib/onewallet.js:
```diff
@@ -451,6 +459,7 @@ module.exports = {
   // commit - specific operations
   computeTransferHash,
   computeRecoveryHash,
+  computeDestHash,
```
merge and export only one

johnwhitton replied:
done

polymorpher replied:
still have both

polymorpher commented on code/package.json:
```diff
@@ -79,7 +80,10 @@
     "eslint-plugin-truffle": "^0.3.1",
     "moment-timezone": "^0.5.33",
     "rlp": "^2.2.6",
+    "solc-0.8.4": "npm:solc@^0.8.4",
```
assume installed on OS

johnwhitton replied:
removed

polymorpher commented on code/package.json:
```diff
@@ -79,7 +80,10 @@
     "rlp": "^2.2.6",
+    "solc-0.8.4": "npm:solc@^0.8.4",
     "solhint": "^3.3.5",
+    "solidity-coverage": "^0.7.20",
```
some dependencies including this should be devDependencies

johnwhitton replied:
moved

polymorpher commented on code/test/util.js:
```diff
@@ -1,25 +1,38 @@
     }
   }
 }
+const ONEDebugger = require('../lib/debug')
```
requires should be on top

johnwhitton replied:
moved

polymorpher commented on code/test/util.js:
```diff
@@ -37,6 +50,47 @@ const deploy = async (initArgs) => {
 
 const ONE_ETH = unit.toWei('1', 'ether')
 
+const sleep = (milliseconds) => {
```
should be marked async

johnwhitton replied:
updated 🙏

polymorpher commented on code/test/util.js:
```diff
@@ -37,6 +50,47 @@ const deploy = async (initArgs) => {
+  Logger.debug(`wait: ${seconds}`)
+  await sleep(seconds * 1000)
+}
+const getReceipt = async (transactionHash) => {
```
waitForReceipt?

johnwhitton replied:
sure, updated 🙏

polymorpher commented on code/test/util.js:
```diff
@@ -215,6 +269,123 @@ const getEOTP = async ({ seed, hseed, effectiveTime, timeOffset }) => {
 }
 
+// transactionExecute commits and reveals a wallet transaction
+const transactionExecute = async ({ wallet, operationType, tokenType, contractAddress, tokenId, dest, amount, data, address, randomSeed, backlinkAddresses, testTime }) => {
```
executeStandardTransaction?

johnwhitton replied:
yes, updated 🙏

polymorpher commented on code/.gitignore:
```diff
@@ -119,6 +119,13 @@ secrets.js
+**/.DS_Store
+
+# Build Files
+build/*
```
just contracts, not abi. Merge master?

johnwhitton replied:
Have merged with master and also updated to 
```
# Build Files
build/contracts
```
🙏

polymorpher commented on code/.env.sample:
```diff
@@ -1,5 +1,16 @@
-ETH_GANACHE_KEY=60537f2b5c5fa6180d3b65c181f005828ddb2799a2d776cf7c5f5ce1935a2f8a
+ETH_GANACHE_MNEMONIC='filter group there hunt fitness junior ghost park route jar entire clown allow rifle meadow'
```
new config from master supports mnemonic and no longer needs multiple keys

johnwhitton replied:
Have removed the keys.
At some point, if needed, we can revisit how we create accounts that can sign.
Currently have left the `ETH_GANACHE_MNEMONIC` in as I use it for testing.
🙏

polymorpher commented on code/test/util.js:
```diff
@@ -215,6 +269,123 @@ const getEOTP = async ({ seed, hseed, effectiveTime, timeOffset }) => {
 
+// transactionExecute commits and reveals a wallet transaction
+const transactionExecute = async ({ wallet, operationType, tokenType, contractAddress, tokenId, dest, amount, data, address, randomSeed, backlinkAddresses, testTime }) => {
+  if (testTime === undefined) { testTime = Date.now() }
```
use default argument

johnwhitton replied:
Great point.
Updated.
🙏

polymorpher commented on code/test/util.js:
```diff
@@ -215,6 +269,123 @@ const getEOTP = async ({ seed, hseed, effectiveTime, timeOffset }) => {
+  const otp = ONEUtil.genOTP({ seed: wallet.seed, counter })
+  // // calculate wallets effectiveTime (creation time) from t0
+  const info = await wallet.wallet.getInfo()
+  const t0 = new BN(info[3]).toNumber()
```
t0 is in wallet

johnwhitton replied:
These two lines are pulling t0 from the wallet
```
  const t0 = new BN(info[3]).toNumber()
  const walletEffectiveTime = t0 * INTERVAL
```
Do you have a recommendation on a more effective/cleaner way to do this?

polymorpher replied:
it is in the walletInfo object you passed in

polymorpher commented on code/test/util.js:
```diff
@@ -215,6 +269,123 @@ const getEOTP = async ({ seed, hseed, effectiveTime, timeOffset }) => {
+  const info = await wallet.wallet.getInfo()
+  const t0 = new BN(info[3]).toNumber()
+  const walletEffectiveTime = t0 * INTERVAL
+  const index = ONEUtil.timeToIndex({ effectiveTime: walletEffectiveTime, time: testTime })
```
effectiveTime also in wallet

johnwhitton replied:
I couldn't find it when searching for effective.
The closest I see is CoreSetting
```
struct CoreSetting {
        /// Some variables can be immutable, but doing so would increase contract size. We are at threshold at the moment (~24KiB) so until we separate the contracts, we will do everything to minimize contract size
        bytes32 root; // Note: @ivan brought up a good point in reducing this to 16-bytes so hash of two consecutive nodes can be done in a single word (to save gas and reduce blockchain clutter). Let's not worry about that for now and re-evalaute this later.
        uint8 height; // including the root. e.g. for a tree with 4 leaves, the height is 3.
        uint8 interval; // otp interval in seconds, default is 30
        uint32 t0; // starting time block (effectiveTime (in ms) / interval)
        uint32 lifespan;  // in number of block (e.g. 1 block per [interval] seconds)
        uint8 maxOperationsPerInterval; // number of transactions permitted per OTP interval. Each transaction shall have a unique nonce. The nonce is auto-incremented within each interval
    }
  ```
  Is there something I missed?

polymorpher replied:
sorry, it wasn't explicitly added in wallet (object you passed in). But if computed from t0, it should also times `interval` (also from the core) and 1000

polymorpher commented on code/test/util.js:
```diff
@@ -215,6 +269,123 @@ const getEOTP = async ({ seed, hseed, effectiveTime, timeOffset }) => {
+  let commitParams
+  let revealParams
+  let paramsHash
+  console.log(`operationType: ${operationType}`)
```
Logger?

johnwhitton replied:
Yes, updated 🙏 
Also by naming the tests more effectively as proposed this will give us information around what is running e.g.
```
it('OPERATION 0 TRACK: must be able to track tokens', async () => {
  })
  ```

polymorpher commented on code/test/util.js:
```diff
@@ -215,6 +269,123 @@ const getEOTP = async ({ seed, hseed, effectiveTime, timeOffset }) => {
+  switch (operationType) {
+    case ONEConstants.OperationType.TRACK:
+    case ONEConstants.OperationType.UNTRACK:
+      paramsHash = ONEWallet.computeGeneralOperationHash
```
remove line 293-296? Params are wrong and there should be only one way of handling computeGeneralOperationHash

johnwhitton replied:
Currently when removing these lines and adding the additional `amount` and `data` fields` and run
`VERBOSE=0 truffle test --network=ganache --compile-none --grep 'OPERATION 0 '`
with the following parameters
```
   testTime = await TestUtil.bumpTestTime(testTime, 60)
    await TestUtil.executeStandardTransaction(
      {
        ...ONEConstants.NullOperationParams, // Default all fields to Null values than override
        wallet: alice,
        operationType: ONEConstants.OperationType.TRACK,
        tokenType: ONEConstants.TokenType.ERC20,
        contractAddress: testerc20.address,
        testTime
      }
    )
```
      
we get
```
     Error: Returned error: VM Exception while processing transaction: revert Param mismatch
      at commitReveal (test/util.js:417:44)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:95:5)
      at Object.executeStandardTransaction (test/util.js:378:3)
      at Context.<anonymous> (test/general.js:53:5)
 ```
This appears to be because of the logic in Executor.sol testing the op.data.length
```
        // No revert should occur below this point
        if (op.operationType == Enums.OperationType.TRACK) {
            if (op.data.length > 0) {
                tokenTrackerState.multiTrack(op.data);
            } else {
                tokenTrackerState.trackToken(op.tokenType, op.contractAddress, op.tokenId);
            }
        } else if (op.operationType == Enums.OperationType.UNTRACK) {
            if (op.data.length > 0) {
                tokenTrackerState.multiUntrack(op.data);
            } else {
                tokenTrackerState.untrackToken(op.tokenType, op.contractAddress, op.tokenId);
            }
```
Hence if we pass data we (incorrectly) call `multitrack` if we omit data by using say the following parameters
```
       //...ONEConstants.NullOperationParams, // Default all fields to Null values than override
        wallet: alice,
        operationType: ONEConstants.OperationType.TRACK,
        tokenType: ONEConstants.TokenType.ERC20,
        contractAddress: testerc20.address,
        TokenId: 1,
        dest: alice.wallet.address,
        amount: 1,
        testTime
        ```
        We get
        ```
             Error: invalid BigNumber value (argument="value", value=undefined, code=INVALID_ARGUMENT, version=bignumber/5.4.1)
      at commitReveal (test/util.js:417:44)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:95:5)
      at Object.executeStandardTransaction (test/util.js:378:3)
      at Context.<anonymous> (test/general.js:53:5)
      ```
      Hence the workaround for now.

polymorpher replied:
Client side implementations (stuff in client/src/pages/Show/*.jsx) all use separate commit params (and hash computing function) and reveal params (and reveal args transformation function, if needed). Here you are attempting to make one function that encompass both commit and reveal functionalities, which have incompatible parameter names, hence the errors / difficulties

polymorpher commented on code/test/util.js:
```diff
@@ -215,6 +269,123 @@ const getEOTP = async ({ seed, hseed, effectiveTime, timeOffset }) => {
+      commitParams = { operationType, tokenType, contractAddress, tokenId, dest, amount, data }
+      revealParams = { operationType, tokenType, contractAddress, tokenId, dest, amount, data }
+      break
+    case ONEConstants.OperationType.RECOVER_SELECTED_TOKENS:
```
dest and data

johnwhitton replied:
Done 🙏 
However this test is still failing. It may have been I just missed something in the test but here's where I got to debugging.
Debugged  it appears to be failing in `Reveal.sol`-> `authenticate`  which calls `authenticateCores`
Wondering whether `RECOVER_SELECTED_TOKENS` should call `isCorrectRecoveryProof` as well
```
        uint32 coreIndex = 0;
        if (op.operationType == Enums.OperationType.RECOVER) {
            coreIndex = isCorrectRecoveryProof(core, oldCores, auth);
        } else {
            coreIndex = isCorrectProof(core, oldCores, auth);
 ```
 Either that or I just missed something in `general.ts` with how we test `RECOVER_SELECTED_TOKENS`

```
Error message
       OPERATION 9 RECOVER_SELECTED_TOKENS: must be able to recover selected tokens:
     Error: Returned error: VM Exception while processing transaction: revert Param mismatch
      at commitReveal (test/util.js:415:44)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:95:5)
      at Object.executeStandardTransaction (test/util.js:376:3)
      at Context.<anonymous> (test/general.js:629:5)
```

polymorpher commented on code/test/util.js:
```diff
@@ -215,6 +269,123 @@ const getEOTP = async ({ seed, hseed, effectiveTime, timeOffset }) => {
+      revealParams = { operationType, tokenType, contractAddress, tokenId, dest, amount, data }
+      break
+    case ONEConstants.OperationType.RECOVER_SELECTED_TOKENS:
+    case ONEConstants.OperationType.BATCH:
```
generalOperation

johnwhitton replied:
Changed 🙏 
TODO: review and update BATCH in general.ts (still failing currently)

polymorpher commented on code/test/util.js:
```diff
@@ -215,6 +269,123 @@ const getEOTP = async ({ seed, hseed, effectiveTime, timeOffset }) => {
 }
 
+// transactionExecute commits and reveals a wallet transaction
+const transactionExecute = async ({ wallet, operationType, tokenType, contractAddress, tokenId, dest, amount, data, address, randomSeed, backlinkAddresses, testTime }) => {
```
backlinkAddresses and randomSeed should not be an argument of a general function?

johnwhitton replied:
Good Point 🙏 
Not sure the best way to do this.
I've added this to a doc to keep track of these items [REFACTOR.md](https://github.com/1wallet-dev/1wallet/blob/testing/code/test/REFACTOR.md)

johnwhitton replied:
If I understand this correctly 
`ONEWallet.sol` calls `Executor.sol` using
`            Executor.execute(op, tokenTrackerState, backlinkAddresses, signatures, spendingState);`
but only receives in the OperationParameters in the reveal call
**The rest come from the ONEWallets state**
e.g.
`    IONEWallet[] backlinkAddresses; // to be set in next version - these are addresses forwarding funds and tokens to this contract AND must have their forwardAddress updated if this contract's forwardAddress is set or updated. One example of such an address is a previous version of the wallet "upgrading" to a new version. See more at https://github.com/polymorpher/one-wallet/issues/78`
So we should only ever pass Operation Paramaeters 
Even in `commitParams` and `revealParams`

polymorpher commented on code/test/util.js:
```diff
@@ -215,6 +269,123 @@ const getEOTP = async ({ seed, hseed, effectiveTime, timeOffset }) => {
 }
 
+// transactionExecute commits and reveals a wallet transaction
+const transactionExecute = async ({ wallet, operationType, tokenType, contractAddress, tokenId, dest, amount, data, address, randomSeed, backlinkAddresses, testTime }) => {
```
I suggest breaking down this function and only  let it handle general, frequently occurring cases. Transaction execution functions for infrequent operations, if needed, can live in their own files. People testing individual functions shouldn't need to come to this centralized function and going over the entire thing to understand what's needed

johnwhitton replied:
Good Point 🙏 
Not sure the best way to do this.
I've added this to a doc to keep track of these items [REFACTOR.md](https://github.com/1wallet-dev/1wallet/blob/testing/code/test/REFACTOR.md)

polymorpher commented on code/test/util.js:
```diff
@@ -215,6 +269,123 @@ const getEOTP = async ({ seed, hseed, effectiveTime, timeOffset }) => {
+      commitParams = { operationType, backlinkAddresses, data }
+      revealParams = { operationType, backlinkAddresses, data }
+      break
+    case ONEConstants.OperationType.COMMAND:
```
this is general operation

johnwhitton replied:
If I understand this correctly 
`ONEWallet.sol` calls `Executor.sol` using
`            Executor.execute(op, tokenTrackerState, backlinkAddresses, signatures, spendingState);`
but only receives in the OperationParameters in the reveal call
**The rest come from the ONEWallets state**
e.g.
`    IONEWallet[] backlinkAddresses; // to be set in next version - these are addresses forwarding funds and tokens to this contract AND must have their forwardAddress updated if this contract's forwardAddress is set or updated. One example of such an address is a previous version of the wallet "upgrading" to a new version. See more at https://github.com/polymorpher/one-wallet/issues/78`
So we should only ever pass Operation Paramaeters 
Even in `commitParams and `revealParams`

polymorpher replied:
The type of operation (i.e. how paramsHash should be computed) is determined by https://github.com/polymorpher/one-wallet/blob/477e6f043e3c956ffe8ee921652575e1630feb26/code/contracts/Reveal.sol#L39 , not what is passed to  executor

polymorpher commented on code/test/util.js:
```diff
@@ -215,6 +269,123 @@ const getEOTP = async ({ seed, hseed, effectiveTime, timeOffset }) => {
+    case ONEConstants.OperationType.BACKLINK_DELETE:
+    case ONEConstants.OperationType.BACKLINK_OVERRIDE:
+      paramsHash = ONEWallet.computeDataHash
+      commitParams = { operationType, backlinkAddresses, data }
```
backlinkAddresses and operationType are not commit or reveal params

polymorpher commented on code/test/util.js:
```diff
@@ -215,6 +269,123 @@ const getEOTP = async ({ seed, hseed, effectiveTime, timeOffset }) => {
+      revealParams = { operationType, backlinkAddresses, data }
+      break
+    case ONEConstants.OperationType.COMMAND:
+      paramsHash = ONEWallet.computeDataHash
```
most commit and reveal params below do not match the hash functions. Some operation types do not match corresponding commit hash functions (should not be called "paramsHash"). This whole function needs some major refactoring

polymorpher commented on code/test/util.js:
```diff
@@ -245,8 +415,8 @@ const commitReveal = async ({ layers, Debugger, index, eotp, paramsHash, commitP
   const wouldSucceed = await wallet.reveal.call(authParams, revealParams)
   Logger.debug(`Reveal success prediction`, !!wouldSucceed)
-  const tx = await wallet.reveal(authParams, revealParams)
-  return { tx, authParams, revealParams }
```
why is tx not returned? Many tests in spendLimit and innerCore rely on tx

johnwhitton replied:
Apologies, I removed when debugging something and forgot to add it back in.
Fixed now 🙏

polymorpher commented on code/test/util.js:
```diff
@@ -257,7 +427,234 @@ const printInnerTrees = ({ Debugger, innerTrees }) => {
 
+// makeWallet uses an index and unlocked web3.eth.account and creates and funds a ONEwallet
+const makeWallet = async (salt, deployer, effectiveTime, duration, setLastResortAddress) => {
+  if (duration === undefined) { duration = DURATION }
```
use default parameters?

johnwhitton replied:
done 🙏

polymorpher commented on code/test/util.js:
```diff
@@ -257,7 +427,234 @@ const printInnerTrees = ({ Debugger, innerTrees }) => {
+    to: wallet.address,
+    value: HALF_ETH
+  })
+  return { wallet, seed, hseed, root, layers, lastResortAddress: lastResortAccount.address }
```
just lastResortAddress

polymorpher replied:
seed, hseed, root, layers are already in wallet

johnwhitton replied:
Thanks for the clarification 🙏 
I have removed `root`
This code was based on `basic.js` and the parameters are used for generating One Time Passwords and calling the transactions
They are used in `executeStandardTransaction` and will need to re replaced before removing.
seed : `const otp = ONEUtil.genOTP({ seed: walletInfo.seed, counter })`
hseed: `const eotp = await ONE.computeEOTP({ otp, hseed: walletInfo.hseed })`
layers : 
```
 let { tx, authParams, revealParams: returnedRevealParams } = await commitReveal({
    Debugger,
    layers: walletInfo.layers,
    index,
    eotp,
    paramsHash,
    commitParams,
    revealParams,
    wallet: walletInfo.wallet
  })
```



If I understand this correctly
ONEWallet.sol calls Executor.sol using
Executor.execute(op, tokenTrackerState, backlinkAddresses, signatures, spendingState);
but only receives in the OperationParameters in the reveal call
The rest come from the ONEWallets state
e.g.
IONEWallet[] backlinkAddresses; // to be set in next version - these are addresses forwarding funds and tokens to this contract AND must have their forwardAddress updated if this contract's forwardAddress is set or updated. One example of such an address is a previous version of the wallet "upgrading" to a new version. See more at https://github.com/polymorpher/one-wallet/issues/78
So we should only ever pass Operation Paramaeters
Even in commitParams and revealParams`

polymorpher commented on code/test/util.js:
```diff
@@ -257,7 +427,234 @@ const printInnerTrees = ({ Debugger, innerTrees }) => {
+  if (duration === undefined) { duration = DURATION }
+  let lastResortAccount = web3.eth.accounts.create()
+  let lastResortAddress = lastResortAccount.address
+  if (setLastResortAddress === false) { lastResortAddress = ONEConstants.EmptyAddress }
```
use default arguments? would be easier to pack them inside an object

polymorpher commented on code/test/util.js:
```diff
@@ -257,7 +427,234 @@ const printInnerTrees = ({ Debugger, innerTrees }) => {
+  await web3.eth.sendTransaction({
+    from: deployer,
+    to: wallet.address,
+    value: HALF_ETH
```
make a parameter?

polymorpher commented on code/test/util.js:
```diff
@@ -257,7 +427,234 @@ const printInnerTrees = ({ Debugger, innerTrees }) => {
+  // create an ERC20
+  const testerc20 = await TestERC20.new(10000000, { from: owner })
+  // create an ERC721
+  const tids = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
```
okay, but could use https://lodash.com/docs/4.17.15#range

johnwhitton replied:
great idea, done 🙏

polymorpher commented on code/test/util.js:
```diff
@@ -257,7 +427,234 @@ const printInnerTrees = ({ Debugger, innerTrees }) => {
+  const testerc20 = await TestERC20.new(10000000, { from: owner })
+  // create an ERC721
+  const tids = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
+  const uris = ['ipfs://test721/0', 'ipfs://test721/1', 'ipfs://test721/2', 'ipfs://test721/3', 'ipfs://test721/4', 'ipfs://test721/5', 'ipfs://test721/6', 'ipfs://test721/7', 'ipfs://test721/8', 'ipfs://test721/9']
```
tids.map(e=> `ipfs://test721/${e}`)

johnwhitton replied:
done 🙏

polymorpher commented on code/test/util.js:
```diff
@@ -257,7 +427,234 @@ const printInnerTrees = ({ Debugger, innerTrees }) => {
+      identificationKeys.push(x[0].toString())
+    }
+  } catch (ex) {
+    console.log(`Failed to parse walletIdentificationKeys: ${ex.toString()}`)
```
error - and this should abort because something is very wrong

johnwhitton replied:
done 🙏

polymorpher commented on code/test/util.js:
```diff
@@ -257,7 +427,234 @@ const printInnerTrees = ({ Debugger, innerTrees }) => {
+  const walletIdentificationKeys = await wallet.getIdentificationKeys()
+  // console.log(`walletIdentificationKeys: ${JSON.stringify(walletIdentificationKeys)}`)
+  let identificationKeys = []
+  try {
```
might be easier to just remove the try block

johnwhitton replied:
done 🙏

polymorpher commented on code/test/util.js:
```diff
@@ -257,7 +427,234 @@ const printInnerTrees = ({ Debugger, innerTrees }) => {
+      extra: new BN(walletInfo[7]).toNumber()
+    }
+  } catch (ex) {
+    console.log(`Failed to parse Wallet Info: ${ex.toString()}`)
```
error and abort

polymorpher commented on code/test/util.js:
```diff
@@ -257,7 +427,234 @@ const printInnerTrees = ({ Debugger, innerTrees }) => {
+  const forwardAddress = (await wallet.getForwardAddress()).toString()
+  const walletInfo = await wallet.getInfo()
+  let info = {}
+  try {
```
remove try? error should cause abortion

johnwhitton replied:
done 🙏

polymorpher commented on code/test/util.js:
```diff
@@ -257,7 +427,234 @@ const printInnerTrees = ({ Debugger, innerTrees }) => {
+  }
+  const walletOldInfo = await wallet.getOldInfos()
+  let oldInfo = []
+  try {
```
remove try?

johnwhitton replied:
done 🙏

polymorpher commented on code/test/util.js:
```diff
@@ -257,7 +427,234 @@ const printInnerTrees = ({ Debugger, innerTrees }) => {
+  }
+  const walletInnerCores = await wallet.getInnerCores()
+  let innerCores = []
+  try {
```
remove try?

johnwhitton replied:
done 🙏

polymorpher commented on code/test/util.js:
```diff
@@ -257,7 +427,234 @@ const printInnerTrees = ({ Debugger, innerTrees }) => {
+  }
+  const nonce = new BN(await wallet.getNonce()).toNumber()
+  const lastOperationTime = new BN(await wallet.lastOperationTime()).toNumber()
+  const walletAllCommits = await wallet.getAllCommits()
```
```
const [hashes, paramsHashes, verificationHashes, timestamps, completed] = Object.keys(walletAllCommits).map(k => walletAllCommits[k])
const commits = hashes.map((e, i) => ({ hash: hashes[i], paramsHash: paramsHashes[i], verificationHash: verificationHashes[i], timestamp: timestamps[i], completed: completed[i] })
```

polymorpher replied:
can be used for all structure below

polymorpher replied:
can be used for all structure below

johnwhitton replied:
Looks very useful. 🙏 
I added in the code and tested and got the error below.
So leaving it in as comments until the error is resolved.

```
     AssertionError: wallet.allCommits is incorrect: expected [ Array(1) ] to deeply equal { Object (0, 1, ...) }
      at Object.checkONEWalletStateChange (test/util.js:615:10)
      at Context.<anonymous> (test/general.js:280:20)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:95:5)
```

polymorpher replied:
that is saying you get two forms of representations (array of objects with named properties, v.s. just a bunch of arrays) , you need to stick with one form everywhere

polymorpher commented on code/test/upgrade.js:
```diff
@@ -0,0 +1,76 @@
+// Test upgrade scenario
```
do later? let's get some stuff merged first

johnwhitton replied:
ok

polymorpher replied:
remove until implemented?

polymorpher commented on code/test/general.js:
```diff
@@ -0,0 +1,2306 @@
+    assert.equal(1000, aliceBalanceERC20, 'Transfer of 1000 ERC20 tokens to alice.wallet succesful')
+    aliceWalletBalanceERC20 = await alice.wallet.getBalance(ONEConstants.TokenType.ERC20, testerc20.address, 0)
+    assert.equal(1000, aliceWalletBalanceERC20, 'Transfer of 1000 ERC20 tokens to alice.wallet checked via wallet succesful')
+    console.log('Funded ERC20')
```
Logger?

johnwhitton replied:
removed :)

polymorpher commented on code/test/general.js:
```diff
@@ -0,0 +1,2306 @@
+    // nonce
+    let nonce = await alice.wallet.getNonce()
+    assert.notEqual(nonce, aliceOldState.nonce, 'alice wallet.nonce should have been changed')
+    assert.equal(nonce.toNumber(), aliceOldState.nonce + 1, 'alice wallet.nonce should have been changed')
```
just need one between 68,69

johnwhitton replied:
Removed 68 and Refactored with a test that accepts 🙏

polymorpher commented on code/test/general.js:
```diff
@@ -0,0 +1,2306 @@
+    // commits
+    let allCommits = await alice.wallet.getAllCommits()
+    assert.notDeepEqual(allCommits, aliceOldState.allCommits, 'alice wallet.allCommits should have been updated')
+    aliceOldState.allCommits = allCommits
```
I think checks for operation time, nonce, commits can be put inside one util function

johnwhitton replied:
Done now in validateUpdateTransaction 🙏

polymorpher commented on code/test/general.js:
```diff
@@ -0,0 +1,2306 @@
+    testTime = await TestUtil.bumpTestTime(testTime, 60)
+
+    // alice tranfers ONE CENT to bob
+    await TestUtil.transactionExecute(
```
Should check for change of recovery address after this

polymorpher commented on code/test/general.js:
```diff
@@ -0,0 +1,2306 @@
+    let randomSeed = new Uint8Array(new BigUint64Array([0n, BigInt(testTime)]).buffer)
+    // recover Alices wallet
+    //
+    await TestUtil.transactionExecute(
```
see basic.js:150. Recovery has special args. Probably shouldn't use this generation execution function

polymorpher replied:
also should check balance of both accounts after this is done

polymorpher commented on code/test/general.js:
```diff
@@ -0,0 +1,2306 @@
+    await TestUtil.checkONEWalletStateChange(aliceOldState, aliceCurrentState)
+  })
+
+  // ==== DISPLACE =====
```
remove? save for later PR and use a separate file

polymorpher replied:
remove tests below that are not implemented / not ready for production or auditing / are boilerplates? Add them later in separate PR when they are production ready

polymorpher commented on code/test/general.js:
```diff
@@ -0,0 +1,2306 @@
+const TestUtil = require('./util')
```
Generally, break tests into small pieces and have each file focus on a single specialized area of features. This is too long and overlaps with many existing tests

polymorpher commented on code/test/tokens.js:
```diff
@@ -0,0 +1,501 @@
+  })
+
+  // Transfer Native Asset to external wallet
+  it('Wallet_CommitReveal: Native Asset Transfer must commit and reveal successfully', async () => {
```
For each test, let's only focus on validating specific feature, rather than everything related to how the wallet should function as of this version - the wallet's interfaces, features, and expected functions may change across versions and over time. When they change, we don't want to update all tests in all files. We only want to modify one test specifically relevant to that feature, and change a small part of the code in a specific file.

Simple transfer is already tested in basic - let's remove it from here and only tests for tokens. Similarly Wallet_Validate can be removed

polymorpher commented on code/test/tokens.js:
```diff
@@ -0,0 +1,501 @@
+/* TODO
```
Let's simplify this file (and maybe other files) and remove as much boilerplates / redundant code as possible, and only test for specific features/results related to tokens. We can have a general health check (or state-change check) function that's called in one line (ideally in before/after hooks, rather than repeated in every test), but let's not have the same code block repeating everywhere

polymorpher commented on code/test/README.md:
```diff
@@ -0,0 +1,113 @@
+# Contributors Testing Guide
+
+## Overview
+Testing is designed to provide complete coverage of all ONEWallet components including the client, relayer and smart contracts.
```
relayer is not yet tested

polymorpher commented on code/test/README.md:
````diff
@@ -0,0 +1,113 @@
+### Running the tests
+It is recommended that you run tests against a local ganache instance. To do this in a separate window run the following command
+```
+ganache --port 7545
````
use code/env

polymorpher commented on code/test/README.md:
```diff
@@ -0,0 +1,113 @@
+| 20 | REVOKE                       | FAIL	  | Base     |
+| 21 | CALL                         | FAIL	  | Base     |
+| 22 | BATCH                        | FAIL	  | Base     |
+| 23 | NOOP                         | N/A	  | N/A      | obsolete 
```
this is for nulloperationparameter

johnwhitton replied:
updated 🙏

polymorpher commented on code/test/README.md:
```diff
@@ -0,0 +1,113 @@
+# Contributors Testing Guide
```
Overall the framework is nice, but some structure might need some revision - some test files are getting too big and there is too much repeated code. We probably don't need too much repeated basic check after each test for specialized operations.  Testing the logics and expected use cases of each operation is much more important than coverage of basic invocation. See innerCore.js and spendLimit.js tests for example. These tests later guided the design of the UI and some subtleties in how effectiveTime should be set and how to find correct nodes in merkle trees. 

Also  when functionalities evolve or change, it may force too much code change.

johnwhitton commented:
I started creating validation helper functions and refactored a number of the functions I wrote to clean up the parameters and take default values.
An example is `validateUpdateTransaction` which simplifies and removes a lot of boilerplate validation logic
https://github.com/1wallet-dev/1wallet/blob/testing/code/test/util.js#L433
I also made a first pass of cleaning up `general.ts` , cleaned up some of the tests and removed over 1300 lines of code
A good idea of what I’m aiming for is the cleaned up TRANSFER test
https://github.com/1wallet-dev/1wallet/blob/testing/code/test/general.js#L250
it has four lines for validation (`validateBalance ` the generic `validateUpdateTransaction`, `validateSpendingState` and `checkONEWalletStateChange`
It also includes a balance check for Bob.
These functions could also be moved into `executeStandardTransaction` if we wanted to reduce boilerPlating further.

polymorpher commented:
code/build/contracts-old/ONEWalletV5.json and V6 and those in `build/abi/` should not be deleted.

polymorpher commented on code/client/src/pages/Show/SetRecovery.jsx:
```diff
@@ -60,7 +60,7 @@ const SetRecovery = ({ address, onClose }) => {
       otp2,
       commitHashGenerator: ONE.computeSetRecoveryAddressHash,
-      commitHashArgs: { address: dest },
+      commitHashArgs: { dest },
```
the hash generator function is not changed

polymorpher commented on code/config.js:
```diff
@@ -25,6 +25,7 @@ module.exports = {
     },
     'eth-ganache': {
       url: process.env.GANACHE_RPC || 'http://127.0.0.1:7545',
+      mnemonic: process.env.ETH_GANACHE_MNEMONIC,
```
repeating below

polymorpher commented on code/.gitignore:
```diff
@@ -119,6 +119,13 @@ secrets.js
+**/.DS_Store
+
+# Build Files
+build/contracts
```
already exists below

polymorpher commented on code/.gitignore:
```diff
@@ -119,6 +119,13 @@ secrets.js
 webclient/dist
 
+# DS_Store files
+**/.DS_Store
```
exists in root .gitignore

polymorpher commented on code/test/util.js:
```diff
@@ -155,64 +280,467 @@ const createWallet = async ({
+// }
+
+// validateUpdateTransaction: changed - nonce, lastOperationTime, commits,
+const validateUpdateTransaction = async ({ wallet, oldState, validateNonce = true }) => {
```
This function should not mutate its parameters. Change the name or break it down to two functions?

polymorpher commented on code/test/util.js:
```diff
@@ -155,64 +280,467 @@ const createWallet = async ({
+}
+
+// validateSpendingState
+const validateSpendingState = async ({ wallet, oldState, spentAmount = 0 }) => {
```
same as above

polymorpher commented on code/test/util.js:
```diff
@@ -155,64 +280,467 @@ const createWallet = async ({
+// const validateTokensTracked = async ({ wallet, walletOldState }) => {
+// }
+
+// ==== STATE RETREIVAL AND VALIDATION FUNCTIONS =====
```
You can also put functions in different objects for grouping

polymorpher commented on code/test/events.js:
```diff
@@ -0,0 +1,172 @@
+// const unit = require('ethjs-unit')
+// const ONEUtil = require('../lib/util')
+const ONEConstants = require('../lib/constants')
+// const BN = require('bn.js')
```
remove unused?

polymorpher commented on code/test/events.js:
```diff
@@ -0,0 +1,172 @@
+// Test all events emitted
```
Cleanup? Also this does not appear to be testing event emissions. Perhaps remove and add in later PRs after it is implemented?

polymorpher commented on code/test/general.js:
```diff
@@ -0,0 +1,945 @@
+const TestUtil = require('./util')
```
Are some of these tests repeating tests that are already in other files (e.g. transfer, token transfer, tracking)? If so, shouldn't they be removed? Also, to avoid more repetitions, and to make scope of testing clear for each test, can all tests in this file be moved to files designated for specific purposes? We can use a .md file to track coverage of each operation, instead of a `general` file that repeats tests already performed in each scope

polymorpher commented:
Merging to `testing` branch for now. Going to start another PR (from `testing` to `master`) to continue the review
